apps/opik-backend/src/main/java/com/comet/opik/domain/ChatCompletionService.java

```diff
@@ -16,170 +14,139 @@
             @NonNull ChatCompletionRequest request, @NonNull String workspaceId) {
         log.info("Creating and streaming chat completions, workspaceId '{}', model '{}'", workspaceId, request.model());
-        var openAiClient = getAndConfigureOpenAiClient(request, workspaceId);
         var chunkedOutput = new ChunkedOutput<String>(String.class, "\r\n");
-        openAiClient.chatCompletion(request)
-                .onPartialResponse(chatCompletionResponse -> send(chatCompletionResponse, chunkedOutput))
-                .onComplete(() -> close(chunkedOutput))
-                .onError(throwable -> handle(throwable, chunkedOutput))
-                .execute();
-        log.info("Created and streaming chat completions, workspaceId '{}', model '{}'", workspaceId, request.model());
+
+        try {
+            var llmProviderClient = llmProviderFactory.getService(workspaceId, request.model());
+
+            llmProviderClient.generateStream(
+                    request,
+                    workspaceId,
+                    getMessageHandler(chunkedOutput),
+                    getCloseHandler(chunkedOutput),
+                    getErrorHandler(chunkedOutput, llmProviderClient));
+            log.info("Created and streaming chat completions, workspaceId '{}', model '{}'", workspaceId,
+                    request.model());
+        } catch (BadRequestException exception) {
+            handleChunckedException(chunkedOutput, new ErrorMessage(HttpStatus.SC_BAD_REQUEST, exception.getMessage()));
+        }
+
         return chunkedOutput;
     }
 
-    private OpenAiClient getAndConfigureOpenAiClient(ChatCompletionRequest request, String workspaceId) {
-        var llmProvider = getLlmProvider(request.model());
-        var encryptedApiKey = getEncryptedApiKey(workspaceId, llmProvider);
-        return newOpenAiClient(encryptedApiKey);
+    private RetryUtils.RetryPolicy newRetryPolicy() {
+        var retryPolicyBuilder = RetryUtils.retryPolicyBuilder();
+        Optional.ofNullable(llmProviderClientConfig.getMaxAttempts()).ifPresent(retryPolicyBuilder::maxAttempts);
+        Optional.ofNullable(llmProviderClientConfig.getJitterScale()).ifPresent(retryPolicyBuilder::jitterScale);
+        Optional.ofNullable(llmProviderClientConfig.getBackoffExp()).ifPresent(retryPolicyBuilder::backoffExp);
+        return retryPolicyBuilder
+                .delayMillis(llmProviderClientConfig.getDelayMillis())
+                .build();
     }
 
-    /**
-     * The agreed requirement is to resolve the LLM provider and its API key based on the model.
-     * Currently, only OPEN AI is supported, so model param is ignored.
-     * No further validation is needed on the model, as it's just forwarded in the OPEN AI request and will be rejected
-     * if not valid.
-     */
-    private LlmProvider getLlmProvider(String model) {
-        return LlmProvider.OPEN_AI;
+    private <T> Consumer<T> getMessageHandler(ChunkedOutput<String> chunkedOutput) {
```

Comment:
Can we move the handling of the response to the controller?

Reply:
Done in https://github.com/comet-ml/opik/pull/957/commits/e0c807683a8bcbe37fae54b48d49a2983ead6390